activerecord/lib/active_record/autosave_association.rb

```diff
@@ -200,13 +200,13 @@ def add_autosave_association_callbacks(reflection)
             after_create save_method
             after_update save_method
           else
-            define_non_cyclic_method(save_method) { save_belongs_to_association(reflection) }
+            define_non_cyclic_method(save_method) { throw(:abort) if save_belongs_to_association(reflection) == false }
             before_save save_method
           end
 
           if reflection.validate? && !method_defined?(validation_method)
             method = (collection ? :validate_collection_association : :validate_single_association)
-            define_non_cyclic_method(validation_method) { send(method, reflection) }
+            define_non_cyclic_method(validation_method) { send(method, reflection) || nil }
```

Comment:
Whenever a callback returns `false`, the user gets a deprecation warning.

In the rest of the code, I have changed all internal uses of `return false` to `throw(:abort)`, so users will not get deprecation warnings for internal code.

However, in this specific case, there is no need to do that because the value returned by this method (whether `nil` or `false`) is not used anywhere. The method above is [invoked here](https://github.com/rails/rails/blob/8b07f18599854475a69c1447642dd3d80bea8df9/activemodel/lib/active_model/validations.rb#L394) inside `_run_validate_callbacks`:

```ruby
    def run_validations! #:nodoc:
      _run_validate_callbacks
      errors.empty?
    end
```

The only important thing is whether `errors` is empty or not.
Since it's not meaningful whether the callback returns `false` or `not`, I added the `|| nil`, so the user will not get a useless deprecation warning. 

---

The following is to show you how you would get deprecation warning when running tests if you didn't add the `|| nil`:

```bash
DEPRECATION WARNING: Returning `false` in a callback will not implicitly halt a callback chain in the next release of Rails. To explicitly halt a callback chain, please use `throw :abort` instead. (called from block in <class:TestAutosaveAssociationValidationsOnABelongsToAssociation> at test/cases/autosave_association_test.rb:1502)
/Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:617:in `display_deprecation_warning_for_false_terminator'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:607:in `block (2 levels) in default_terminator'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:604:in `catch'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:604:in `block in default_terminator'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:166:in `call'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:166:in `block in halting'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:506:in `call'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:506:in `block in call'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:506:in `each'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:506:in `call'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:93:in `_run_callbacks'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:808:in `_run_validate_callbacks'
  /Users/claudiob/code/rails/activemodel/lib/active_model/validations.rb:397:in `run_validations!'
  /Users/claudiob/code/rails/activemodel/lib/active_model/validations/callbacks.rb:112:in `block in run_validations!'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:89:in `call'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:89:in `_run_callbacks'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:808:in `_run_validation_callbacks'
  /Users/claudiob/code/rails/activemodel/lib/active_model/validations/callbacks.rb:112:in `run_validations!'
  /Users/claudiob/code/rails/activemodel/lib/active_model/validations.rb:334:in `valid?'
  /Users/claudiob/code/rails/activerecord/lib/active_record/validations.rb:58:in `valid?'
  test/cases/autosave_association_test.rb:1502:in `block in <class:TestAutosaveAssociationValidationsOnABelongsToAssociation>'
```